lib/daemon_runner/session.rb

```diff
@@ -0,0 +1,124 @@
+  #
+  class Session
+    include Logger
+
+    class << self
+      attr_reader :session
+
+      def start(name, **options)
+        @session ||= Session.new(name, options).renew!
+      end
+
+
+      # Acquire a lock with the current session, or initialize a new session
+      #
+      # @param path [String]  A path in the Consul key-value space to lock
+      # @param lock_session [Session] The Session instance to lock the lock to
+      # @return [Boolean] `true` if the lock was acquired
+      #
+      def lock(path)
+        Diplomat::Lock.acquire(path, session.id)
+      end
+
+      # Release a lock held by the current session
+      #
+      # @param path [String]  A path in the Consul key-value space to release
+      # @param lock_session [Session] The Session instance that the lock was acquired with
+      #
+      def release(path)
+        Diplomat::Lock.release(path, session.id)
+      end
+    end
+
+    # Consul session ID
+    attr_reader :id
+
+    # Session name
+    attr_reader :name
+
+    # Period, in seconds, after which session expires
+    attr_reader :ttl
+
+    # Period, in seconds, that a session's locks will be
+    attr_reader :delay
+
+    attr_reader :behavior
+
+    # @param name [String]  Session name
+    # @option options [Fixnum]  ttl (15)    Session TTL in seconds
+    # @option options [Fixnum]  delay (15)  Session release dealy in seconds
+    # @option options [String]  behavior (release)  Session release behavior
+    def initialize(name, **options)
+      logger.info('Initializing a Consul session')
+
+      @name = name
+      @ttl = options.fetch(:ttl, 15)
+      @delay = options.fetch(:delay, 15)
+      @behavior = options.fetch(:behavior, 'release')
+
+      init
+    end
+
+    # Check if there is an active renew thread
+    #
+    # @return [Boolean] `true` if the thread is alive
+    def renew?
+      @renew.is_a?(Thread) && @renew.alive?
+    end
+
+    # Create a thread to periodically renew the lock session
+    #
+    def renew!
+      return if renew?
+
+      @renew = Thread.new do
+        ## Wakeup every TTL/2 seconds and renew the session
+        loop do
+          sleep ttl / 2
+
+          begin
+            logger.debug(" - Renewing Consul session #{id}")
+            Diplomat::Session.renew(id)
+
+          rescue Faraday::ResourceNotFound
+            logger.warn("Consul session #{id} has expired!")
+
+            init
+          rescue StandardError => e
+            ## Keep the thread from exiting
+            logger.error(e)
+          end
+        end
+      end
+
+      self
+    end
+
+    # Stop the renew thread and destroy the session
+    #
+    def destroy!
+      @renew.kill if renew?
+      Diplomat::Session.destroy(id)
+    end
+
+    private
+
+    # Initialize a session and store it's ID
+    #
+    def init
+      @id = Diplomat::Session.create(
+        :Name => name,
+        :TTL => "#{ttl}s",
+        :LockDelay => "#{delay}s",
+        :Behavior => behavior
```

Comment:
I get what you're doing here, but what's the point of accessing these variables via the `:attr_reader` rather than the corresponding instance variable? Imo it makes more sense to just use `@name`, `@ttl`, `@delay`, and `@behavior` as it's a bit more explicit where the variables are coming from.